client/src/unifyfs-sysio.c

```diff
@@ -338,50 +338,45 @@ int UNIFYFS_WRAP(remove)(const char* path)
 /* The main stat call for all the *stat() functions */
 static int __stat(const char* path, struct stat* buf)
 {
-    int gfid, fid;
+    int gfid, fid, ret;
     unifyfs_file_attr_t fattr;
-    int ret;
-
-    gfid = unifyfs_generate_gfid(path);
-    fid = unifyfs_get_fid_from_path(path);
 
     /* check that caller gave us a buffer to write to */
     if (!buf) {
         errno = EFAULT;
         return -1;
     }
+    memset(buf, 0, sizeof(*buf));
 
     /* lookup stat data for global file id */
-    ret = invoke_client_metaget_rpc(gfid, &fattr);
+    gfid = unifyfs_generate_gfid(path);
+    fid = unifyfs_get_fid_from_path(path);
```

Comment:
This PR does not introduce the issue, but what happens in the case that a process stats a file that it has never opened?  I think it would not have a valid fid, so this will likely return -1.  I think that process may then see an invalid global file size.

Reply:
And an issue the writeread case brings up which this is trying to fix.  I guess we haven't defined or documented when or how a process can detect that a file it already has open has been laminated.

Reply:
if the fid is -1 (i.e., no local metadata), then we don't try to update the local metadata.

there were a couple of other places that I fixed where we were assuming a good fid was being passed in a call to get the local metadata, and the resultant (possibly NULL) pointer was being dereferenced.

Reply:
you're right about the undefined nature of discovering lamination status. with current code, you have to laminate (chmod from rank0), barrier, then stat() from other ranks to ensure a consistent view, which is what I modified our testutil `write_laminate()` to do.

I suppose we could punt for now and say only the rank that does the laminate knows the status, and it's up to application-level logic to ensure a consistent view. the alternative requires us to push the new status to processes that have it open, which I can't think of a reasonable way to accomplish.

Reply:
Sounds good.  Thanks, @MichaelBrim .  That also pairs up well as the reverse of this function which sets the global file metadata based on the local metadata for a fid:
https://github.com/LLNL/UnifyFS/blob/66448e0d791362bc303a31a07537d03c5a1c3635/client/src/unifyfs-internal.h#L564